File: SOLID/OCP/open-closed-principle.go
package main

import "fmt"


type PaymentMethod interface{
	Pay(amount int)
}

type BkashPayment struct{}

func(b BkashPayment) Pay(amount int){
	fmt.Printf("Paying through Bkash of Tk: %d\n", amount)
}
type DebitCardPayment struct{}
func(dc DebitCardPayment) Pay(amount int){
	fmt.Printf("Paying through DebitCard of Tk: %d\n", amount)
}
type PayPalPayment struct{}
func(dc PayPalPayment) Pay(amount int){
	fmt.Printf("Paying through PayPal of Tk: %d\n", amount)
}
type PaymentProcessor struct{}
func(pp PaymentProcessor) ProcessPayment(paymentMethod PaymentMethod,amount int){
	paymentMethod.Pay(amount)
}
func main(){
	bkash:=BkashPayment{}
	debit := DebitCardPayment{}

	paypal := PayPalPayment{}
	paymentProcessor := PaymentProcessor{}
	paymentProcessor.ProcessPayment(bkash, 500)
	paymentProcessor.ProcessPayment(debit, 1000)
	paymentProcessor.ProcessPayment(paypal, 700)
}

File: SOLID/OCP/open-closed-principle.py
#Open Closed Principle
from abc import ABC,abstractmethod

class PaymentMethod(ABC):
    @abstractmethod
    def pay(self,amount:int):
        pass
class BkashPayment(PaymentMethod):
    def pay(self, amount:int):
        print(f"Paring through Bkash of Tk: {amount}")

class DebitCardPayment(PaymentMethod):
    def pay(self,amount:int):
        print(f"Paring through DebitCard of Tk: {amount}")
class PayPalPayment(PaymentMethod):
    def pay(self,amount:int):
        print(f"Paring through PayPal of Tk: {amount}")
class PaymentProcessor:
    def process_payment(self,payment_method:PaymentMethod,amount:int):
        payment_method.pay(amount)
bKash=BkashPayment()
debit=DebitCardPayment()
paypal=PayPalPayment()

pay_process=PaymentProcessor()
pay_process.process_payment(bKash,500)
pay_process.process_payment(debit,1000)
pay_process.process_payment(paypal,700)
